app/validators/placeholders/placeholder_validator.py

```diff
@@ -108,6 +108,46 @@ def validate_option_label_from_value_placeholder(self, answer_id):
                 identifier=answer_id,
             )
 
+    def validate_answer_type_for_transform(
+        self, argument, argument_name, transform_type
+    ):
+        if (
+            transform_type in ["format_unit", "format_percentage"]
+            and argument_name == "value"
+            and argument.get("source") == "answers"
+        ):
+            answer_id = argument.get("identifier")
+            answer_type = self.questionnaire_schema.answers_with_context[answer_id][
+                "answer"
+            ]["type"]
+            if answer_type.lower() not in transform_type:
+                self.add_error(
+                    error_messages.ANSWER_TYPE_FOR_TRANSFORM_TYPE_INVALID.format(
+                        transform=transform_type, answer_type=answer_type
+                    ),
+                    identifier=answer_id,
+                )
+
+    def validate_answer_and_transform_unit_match(self, arguments, transform_type):
+        if transform_type == "format_unit":
+            answer_id = arguments["value"].get("identifier")
+            unit = arguments["unit"]
+            if (
+                unit
+                != self.questionnaire_schema.answers_with_context[answer_id]["answer"][
+                    "unit"
+                ]
+            ):
+                self.add_error(
+                    error_messages.ANSWER_UNIT_AND_TRANSFORM_UNIT_MISMATCH.format(
+                        answer_unit=self.questionnaire_schema.answers_with_context[
+                            answer_id
+                        ]["answer"]["unit"],
+                        transform_unit=unit,
+                    ),
+                    identifier=answer_id,
+                )
+
     def validate_placeholder_transforms(self, transforms):
```

Comment:
I think we need to clean up this method, there are duplicate validation calls which we should avoid. We should look at doing something like:


```py
    def _validate_placeholder_previous_transforms(self, transforms):
        # First transform can't reference a previous transform
        first_transform = transforms[0]
        for argument_name in first_transform.get("arguments"):
            argument = first_transform["arguments"][argument_name]
            if (
                isinstance(argument, dict)
                and argument.get("source") == "previous_transform"
            ):
                self.add_error(self.FIRST_TRANSFORM_CONTAINS_PREVIOUS_TRANSFORM_REF)

        # Previous transform must be referenced in all subsequent transforms
        for transform in transforms[1:]:
            previous_transform_used = False
            for argument_name in transform.get("arguments"):
                argument = transform["arguments"][argument_name]
                if (
                    isinstance(argument, dict)
                    and argument.get("source") == "previous_transform"
                ):
                    previous_transform_used = True

            if not previous_transform_used:
                self.add_error(self.NO_PREVIOUS_TRANSFORM_REF_IN_CHAIN)

    def validate_placeholder_transforms(self, transforms):
        self._validate_placeholder_previous_transforms(transforms)

        for transform in transforms:
            for argument_name in transform.get("arguments"):
                argument = transform["arguments"][argument_name]
                if (
                    transform["transform"] == "option_label_from_value"
                    and argument_name == "answer_id"
                ):
                    self.validate_option_label_from_value_placeholder(argument)

                self.validate_answer_type_for_transform(
                    transform.get("arguments"), argument_name, transform["transform"]
                )

            self.validate_answer_and_transform_unit_match(
                transform.get("arguments"), transform["transform"]
            )
```

The current methods could do some work to make it consistent with what we have, existing check transform before calling, and the new ones call and then check the transform. However, won't block on this as it is minor.

Reply:
Added, great suggestion, thanks!

Reply:
Done seem to be done.

Reply:
For some reason I haven't pushed it, don't know what happened. Now it is there, thanks!